crates/driver/src/driver.rs

```diff
@@ -101,3 +155,281 @@ impl Driver {
         .map(|receipt| receipt.transaction_hash)
     }
 }
+
+/// Polls the `producer` for new work and buffers only the most recent one. Whenever a
+/// computational task finishes a new one will be started with the most recently buffered work.
+/// Returns the most recent output from a computational task when the `deadline` has been reached.
+async fn last_completed<T, W, B, P, D>(build_task: B, producer: P, deadline: D) -> Option<T>
+where
+    B: Fn(W) -> Pin<Box<dyn FusedFuture<Output = T> + Send>>,
+    P: FusedStream<Item = W>,
+    D: FusedFuture<Output = ()>,
+    T: Send,
+    W: Send,
+{
+    futures::pin_mut!(producer);
+    futures::pin_mut!(deadline);
+
+    let mut result = None;
+    let mut next_work = None;
+
+    let create_pending_future = || {
+        Box::pin(futures::future::pending::<T>().fuse())
+            as Pin<Box<dyn FusedFuture<Output = _> + Send>>
+    };
+
+    let mut currently_computing = false;
+    // initialize with future that does nothing and can be dropped safely
+    let mut current_task = create_pending_future();
```

Comment:
Doesn't this need to initialize with `build_task` right away? Otherwise won't we wait until the next block to start?

Reply:
Maybe you need to pass in a `seed: W` for starting the first task.

Reply:
The [stream](https://github.com/cowprotocol/services/blob/main/crates/shared/src/current_block.rs#L76-L78) in use is a [`tokio_stream::wrappers::WatchStream`](https://docs.rs/tokio-stream/latest/tokio_stream/wrappers/struct.WatchStream.html) which starts with immediately yielding the current value at the time of creation.
Technically we could observe another block between creating the stream and starting to poll it which would result one wasted computation but this time window is extremely small to highly unlikely that we run into that edge case.

Reply:
> out-dated computation but this time window is extremely small so highly unlikely that we run into that edge case.

I think this kind of race-condition is unavoidable.

> which starts with immediately yielding the current value at the time of creation.

Can we add a comment that this is the case? It wasn't immediately clear to me, but you are right that we don't need the "seed" thing I suggested ☝️